doc/developer/design/20230306_persist_mfp_pushdown.md

```diff
@@ -0,0 +1,272 @@
+
+To evaluate an MFP for a particular part in the persist source, we:
+- Calculate the `ResultSpec` for each column using our stats.
+- Evaluate the MFP using the machinery above, returning a `ResultSpec` that captures the possible outputs of the filter.
+- If the filter might return `true` or an error, we keep the part; otherwise, we filter it out.
+
+## Filtering JSONB
+
+The above is enough for normal temporal filters.
+However, we'd also like to be able to push down queries against nested structures, like JSON.
+To be able to tell whether a query like `select * from x where (json_field -> 'ts')::numeric < mz_now()` matches our part,
+we'll need to know the range of possible numeric values in our `'ts'` column.
+(Or whether there could be any non-numeric values that might make the cast fail.)
+
+```rust
+use std::collections::BTreeMap;
+
+pub struct ResultSpec<'a> {
+    // For map datums, constraints on individual fields in the map.
+    fields: BTreeMap<Datum<'a>, ResultSpec<'a>>,
+    // ...and other fields as above: `nullable`, `fallible`, `values`.
+}
+```
+
+This is straightforward to populate from the JSON stats mentioned above.
+
+Finally, in our interpreter, we'll need to fetch any field-specific info from the map when we encounter a field-access function like `JsonbGetString`.
+Since there are a small number of these functions, and they tend to be called with constants on the right-hand side,
+matching on these functions appears to be fairly straightforward.
+
+This composes well with the more general approach to propagating range information described in the previous section.
+In the `json_field` example above, assuming `ts` is always numeric in our shard, our interpreter will:
+- Match on the innermost-nested expression, the field lookup, and extract the nested stats for our field.
+- Apply the cast. Since the cast is monotonic, we'll preserve our range information.
+- `<` will compare two ranges, the one computed in the previous step and the known range for `mz_now`,
+  and determine the possible booleans for the overall expression.
+
+As mentioned, due to stats trimming, we may not have statistics for any particular field.
+To be conservative, we'll assume that that field might have any possible value.
+
+# Rollout
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+Because Persist is so core to Materialize, we get a large amount of test coverage from existing tests. We can additionally write:
+* Proptests to verify that all column types that we support statistics on sort equivalently to their Datum counterparts.
+* Benchmarks to verify the new format's performance in reads and writes.
+* Tests to verify that the new expression interpreter is consistent with `eval`.
```

Comment:
curious what the plan here is

Reply:
This is a ripe area for concrete tests (given a concrete input `e`, if `eval(e) = v` and we abstract `e` into `abs(e)` and `v` into `abs(v)`, then we should have `abs_eval(abs(e)) = abs(v)`) and proptests (don't select `e`, randomly generate it).

We can also test the particular abstract domains by generating values and making sure that `abs(e).may_contain(e)`.

Reply:
Right! And we test both those things... though for the former we check the slightly weaker property `abs_eval(abs(e)).may_contain(eval(e))`. (ie. The result of abstractly interpreting the expression may contain the true result.) I added some language about this to the doc.